Added ability to select search engine

matthewauld opened the pull request:
#1722  Added the ability to select from a list (right now just google.com and duckduckgo.com,) or your own, custom search url.

A few notes:
- in the settings menu, the customs fields don't do any sort of validation. You need to enter a valid url. I am not sure if there is a standard URL regex we are using or not - I can add that in.
- urls.js `examineLocationInput` returns a vSearch result, which I changed just to return the query string to search (eg. `?q=query1+query2`) as opposed to the full url. It is only used in the location bar code as far as I can tell, so it should be fine, as I changed the location bar to work with that.
- this only effects the location bar search. If everyone is happy with this pull, I can hunt down the others and change them to use the user setting instead the hardcoded duckduckgo. I think there is a search option in some of the right click context menus.

Please let me know if you have any feedback!

pfrazee commented on app/bg/dbs/settings.js:
```diff
@@ -39,7 +40,14 @@ export const setup = function (opts) {
     dat_bandwidth_limit_up: 0,
-    dat_bandwidth_limit_down: 0
+    dat_bandwidth_limit_down: 0,
+    selected_search_engine: 0,
+    custom_search_engine_name: '',
+    custom_search_engine_url: '',
+    default_search_engines: [
+        {name:'DuckDuckGo',url:'https://www.duckduckgo.com/'},
+        {name:'Google',url:'https://www.google.com/search'}
+      ]
```
I think we can modify these settings to be a little more robust. Rather than making the custom search engine separate, we can just have it be a part of the list. So:

```
selected_search_engine: 0,
search_engines: [
  {name: 'DuckDuckGo', url: 'https://www.duckduckgo.com/'},
  {name: 'Google', url: 'https://www.google.com/search'}
]
```

Then the user adds a custom engine by adding to that list. A few notes about this:

 1. The settings-db only saves strings and numbers, so we'll need to update the get/set to parse/stringify as JSON for the `search_engines` value. You can just hard code that as a conditional.
 2. Check your tab spacing there, the default_search_engines array is too indented.
 3. Also please put spaces after commas and colons in objects!

matthewauld replied:
Ahh, just store the whole JSON string. Right, that is way easier.

pfrazee commented on app/fg/location-bar/index.js:
```diff
@@ -290,16 +289,25 @@ class LocationBar extends LitElement {
+    var settings = await bg.beakerBrowser.getSettings()
+    var searchEngine
+    //get search engine to use
+    if (settings.selected_search_engine == -1) {
```
The -1 logic here won't be needed after the custom search-engine gets stored to the `search_engines` array, but it _would_ be smart to have DDG as a builtin backup in case the value is bad for some reason.

pfrazee commented on app/userland/settings/js/views/general.js:
```diff
@@ -59,10 +58,64 @@ class GeneralSettingsView extends LitElement {
+            ${this.settings.default_search_engines.map((engine,i)=>{
+
+              return html`
+                        <div class="radio-item">
+                          <input type="radio"
+                                 id="engine${i}"
+                                 name="search-engine"
+                                 value="${i}"
+                                 ?checked="${this.settings.selected_search_engine === i}"
+                                 @change="${this.onSearchEngineChange}"/>
+                          <label for="engine${i}">
+                            ${engine.name} (${engine.url})
+                          </label>
+                        </div>`
```
Tab depth is a little off on this. I'm okay with how you aligned the attributes but the markup chunks should be only 2 spaces deeper than the `return` statement

matthewauld commented:
Works for me! Will make changes.

matthewauld commented:
Alright, updated - stored as a JSON string so user can add as many custom search engines as they want. Fixed the tabbing issues (I think, let me know if I didn't).  Also added logic so they can' t remove the last/current search engine. Custom search engine also affects the context menu (right-click).  Made small changes to css so that `input[type="url"]` has the same styling as `input[type="text"]`

Thanks again for feedback! This is my first contribution to an open source project, so apologies if I miss something.

pfrazee commented on app/bg/dbs/settings.js:
```diff
@@ -44,7 +45,12 @@ export const setup = async function (opts) {
-    dat_bandwidth_limit_down: 0
+    dat_bandwidth_limit_down: 0,
+    selected_search_engine: 0,
+    default_search_engines: [
```
We should probably call this `search_engines` instead of `default_search_engines` since it's user-modifiable

pfrazee commented on app/bg/dbs/settings.js:
```diff
@@ -57,10 +63,12 @@ export const once = events.once.bind(events)
  * @returns {Promise<void>}
  */
 export async function set (key, value) {
+  if (['default_search_engines', 'selected_search_engine'].includes(key))
```
Does `selected_search_engine` need to be JSONified? It's a number, right? Only objects need to be serialized.

matthewauld replied:
Right, Sqlite allows different datatypes in a column... I thought they would all have to be/would be converted to strings.

pfrazee commented on app/bg/dbs/settings.js:
```diff
@@ -57,10 +63,12 @@ export const once = events.once.bind(events)
  * @returns {Promise<void>}
  */
 export async function set (key, value) {
+  if (['default_search_engines', 'selected_search_engine'].includes(key))
```
On conditionals, please use brackets unless it's all one line, so either:

```
if (cond) dothing() // <-- only do this rarely too
```

or

```
if (cond) {
  doThings()
}
```

but never

```
if (cond)
  doThing()
```

We follow this rule to avoid accidental dangling ifs like Heartbleed suffered from.

matthewauld replied:
Makes total sense.

pfrazee commented on app/bg/dbs/settings.js:
```diff
@@ -57,10 +63,12 @@ export const once = events.once.bind(events)
  * @returns {Promise<void>}
  */
 export async function set (key, value) {
+  if (['default_search_engines', 'selected_search_engine'].includes(key))
```
It's a good idea to move the list of keys that need to be JSONified to a constant at the top of the file, something like `JSON_ENCODED_SETTINGS` so that the lists dont fall out of sync in the various parts of the code

pfrazee commented on app/bg/dbs/settings.js:
```diff
@@ -107,9 +119,12 @@ export const getAll = function () {
       rows.forEach(row => { obj[row.key] = row.value })
+      // parse non-string values
+      if (obj.default_search_engines) obj.default_search_engines = JSON.parse(obj.default_search_engines)
+      if (obj.selected_search_engine) obj.selected_search_engine = JSON.parse(obj.selected_search_engine)
```
Make sure to use the `JSON_ENCODED_SETTINGS.includes(key)` logic here instead of hardcoding them.

pfrazee replied:
Same story, wrap in try/catch

pfrazee commented on app/bg/dbs/settings.js:
```diff
@@ -93,7 +101,11 @@ export const get = function (key) {
+      if (row) {
+        row = row.value
+        if (['default_search_engines', 'selected_search_engine'].includes(key))
+          row = JSON.parse(row)
```
We should wrap the `parse()` calls in a try/catch and then fall back to the default value on failure

pfrazee commented on app/fg/location-bar/index.js:
```diff
@@ -290,16 +289,21 @@ class LocationBar extends LitElement {
     var queryId = ++this.queryIdCounter
     this.inputValue = this.inputValue.trim()
     var finalResults
+    var settings = await bg.beakerBrowser.getSettings()
```
We dont want to call `getSettings()` here because it gets called multiple times while the user types and needs to be quick. It should happen in `reset()` and the settings can be stored on the `this`

matthewauld replied:
Sounds good. Since `reset` is not async and `getSettings` is, I changed it so reset calls `this.settingsPromise = bg.beakerBrowser.getSettings()` and `queryAutocomplete` calls `var settings = await this.settingsPromise`. It works, but I am not sure if there is a better/proper way?

pfrazee commented on app/userland/settings/js/views/general.js:
```diff
@@ -484,6 +545,35 @@ class GeneralSettingsView extends LitElement {
+
+  removeSearchEngine (i) {
+    // decrement selected search engine so it points to the same one if the removed index is less than current one
+    if (i<this.settings.selected_search_engine){
```
Sorry to nitpick but can you put spaces around the `<`?

matthewauld replied:
No worries - thought I got them all.

pfrazee commented on app/userland/settings/js/views/general.js:
```diff
@@ -484,6 +545,35 @@ class GeneralSettingsView extends LitElement {
+    const newValue = e.target.value
+    this.settings.selected_search_engine = parseInt(newValue)
+    beaker.browser.setSetting('selected_search_engine', this.settings.selected_search_engine)
+    this.requestUpdate()
```
Can you render a "Settings Updated" toast here? (See the other update handlers for how)

matthewauld replied:
Sure!

matthewauld commented:
Alright - besides being unsure of where I called getSettings in the location-bar (It works, just not sure if it is the right form) otherwise, should be good to go! Sorry for all the back and forth, appreciate the notes.

pfrazee commented:
Okay great work! Testing now

pfrazee commented:
Okay merged! Appreciate you working with me through it. I made a couple of final tweaks before merging (https://github.com/beakerbrowser/beaker/commit/d08989861200bfee44799166e825cc7de5fac585)

matthewauld commented:
Ah - yeah that does work better!
